enhance serverless functions intro (#1402)
zeit/docs: pages/docs/v2/serverless-functions/introduction.mdx

  title: 'Serverless Functions',
  description: 'Learn how to deploy Serverless Functions with ZEIT Now.',
  editUrl: 'pages/docs/v2/serverless-functions/introduction.mdx',
  lastEdited: '2019-10-21T06:49:56.000Z'
}

ZEIT Now deployments support various backend languages that you can deploy as Serverless Functions.

Serverless Functions are pieces of code that run on demand by reacting to events, such as an HTTP request.

> They open the door to an entire world of scalable backend micro services that augment your static website.

Reviewer: I think we can do better here. The second paragraph sounds like a marketing page, whereas we are trying to be analytical and technical in our documentation. 

We could probably merge this into one paragraph also.

> With ZEIT Now, you can deploy Serverless Functions, which are pieces of code written with backend languages that take an HTTP request and provide a response. 

Just an example ^

Reviewer: We may be able to link to wikipedia for backend languages, also
https://en.wikipedia.org/wiki/Front_and_back_ends#Back-end_focused

Author: @timothyis I see why you're saying that it sounds a bit markety, but if its a new concept to people coming from a traditional stack with no idea of serverless or what it can do, don't you think we should mention a few things its suited well for? 

Also, my fear to linking to that wiki page is that if people see a runtime that we don't support, they'll open up unnecessary support tickets or request on how to ship Elixir functions or something.

Author: The other thing regarding serverless functions is that they don't only respond to HTTP requests, but can very well react to different events and services. We've discussed expanding the range of things that can trigger function invocations in the future when we polish that side of the product, so I don't think we should limit it to an HTTP request / response in the description.

Reviewer: > but if its a new concept to people coming from a traditional stack with no idea of serverless or what it can do, don't you think we should mention a few things its suited well for

Yes, but you do that below with use cases. These are just buzzwords which do not fit either the tone nor language of our documentation.

> The other thing regarding serverless functions is that they don't only respond to HTTP requests, but can very well react to different events and services. We've discussed expanding the range of things that can trigger function invocations in the future when we polish that side of the product, so I don't think we should limit it to an HTTP request / response in the description.

This, I think, opens us up for confusion. We currently don't support these other events or services, right? If so, we should mentioned which events and services. Otherwise we'll get question about it for sure.

You can use Serverless Functions to handle user authentication, form submission, database queries, custom

> slack commands and more. For a conceptual overview of this new compute model, click [here](https://zeit.co/docs/v2/advanced/concepts/serverless-functions/#conceptual-model).

Reviewer: Oxford comma
```suggestion
slack commands, and more. For a conceptual overview of this new compute model, click [here](https://zeit.co/docs/v2/advanced/concepts/serverless-functions/#conceptual-model).
```

  title: 'Serverless Functions',
  description: 'Learn how to deploy Serverless Functions with ZEIT Now.',
  editUrl: 'pages/docs/v2/serverless-functions/introduction.mdx',
  lastEdited: '2019-10-22T16:10:50.000Z'
}

With ZEIT Now, you can deploy Serverless Functions, which are pieces of code written with [backend languages](/docs/v2/serverless-functions/supported-languages/#supported-languages) that take an HTTP request and provide a response.

You can use Serverless Functions to handle user authentication, form submission, database queries, custom
slack commands, and more.

## Creating Serverless Functions

> To create a serverless function, create a file in an `api` directory from your project root with the appropriate language extension. For example, to create a TypeScript function, you can create a file called `get-user.ts` like below.

Reviewer 2: ```suggestion
To create a Serverless Function, create a file in an `api` directory from your project root with the appropriate language extension. For example, to create a TypeScript function, you can create a file called `get-user.ts` like below.
```

```ts
import { NowRequest, NowResponse } from '@now/node'

export default (req: NowRequest, res: NowResponse) => {
  res.json({ name: 'John', email: 'john@example.com' })
}
```

> 

Reviewer 2: Caption.

> When a user visits <https://example.com/api/get-user>, they will recieve the json from above.

Reviewer 2: ```suggestion
When a user visits the `/api/get-user` endpoint, they will receive the JSON from above.
```

  title: 'Serverless Functions',
  description: 'Learn how to deploy Serverless Functions with ZEIT Now.',
  editUrl: 'pages/docs/v2/serverless-functions/introduction.mdx',
  lastEdited: '2019-10-22T22:13:11.000Z'
}

With ZEIT Now, you can deploy Serverless Functions, which are pieces of code written with [backend languages](/docs/v2/serverless-functions/supported-languages/#supported-languages) that take an HTTP request and provide a response.

You can use Serverless Functions to handle user authentication, form submission, database queries, custom
slack commands, and more.

## Creating Serverless Functions

To create a Serverless Function, create a file in an `api` directory from your project root with the appropriate language extension. For example, to create a TypeScript function, you can create a file called `get-user.ts` like below.

```ts
import { NowRequest, NowResponse } from '@now/node'

export default (req: NowRequest, res: NowResponse) => {
  res.json({ name: 'John', email: 'john@example.com' })
}
```

<Caption>
  A TypeScript Serverless Function that returns a JSON response.
</Caption>

> When a user visits `/api/get-user`, they will recieve the json from above. Note that we don't have to explicitly set up any routing rules for this to work. By default, [filesystem routing](https://zeit.co/docs/configuration/#routes) is leveraged.

Reviewer 2: ```suggestion
When a user visits `/api/get-user`, they will receive the JSON from above. Note that we don't have to explicitly set up any routing rules for this to work. By default, [filesystem routing](https://zeit.co/docs/configuration/#routes) is leveraged.
```

> When a user visits `/api/get-user`, they will receive the JSON from above. Note that we don't have to explicitly set up any routing rules for this to work. By default, [filesystem routing](https://zeit.co/docs/configuration/#routes) is leveraged.

Reviewer 2: ```suggestion
When a user visits `/api/get-user`, they will receive the JSON from above. Note that you don't have to explicitly set up any routing rules for this to work. By default, [filesystem routing](https://zeit.co/docs/configuration/#routes) is leveraged.
```

  title: 'Serverless Functions',
  description: 'Learn how to deploy Serverless Functions with ZEIT Now.',
  editUrl: 'pages/docs/v2/serverless-functions/introduction.mdx',
  lastEdited: '2019-10-18T05:16:28.000Z'
}

ZEIT Now deployments support various backend languages that you can deploy as Serverless functions.

> Serverless functions are pieces of code that run on demand by reacting to events, such as an HTTP request.

Reviewer 2: ```suggestion
Serverless Functions are pieces of code that run on demand by reacting to events, such as an HTTP request.
```

They open the door to an entire world of scalable backend micro services that augment your static website.

> You can use serverless functions to handle user authentication, form handling, dynamic pdf generation, custom

Reviewer 2: ```suggestion
You can use Serverless Functions to handle user authentication, form submission, database queries, custom
```

Author: Great call w/ the db queries 👍

> slack commands, and more. For a conceptual overview of this new compute model, click [here](https://zeit.co/docs/v2/advanced/concepts/serverless-functions/#conceptual-model).

Reviewer 2: ```suggestion
Slack commands, and more.
```

I've removed the link here as this section is intended to replace the conceptual overview.

Author: I capitalized the Slack, but can we talk about removal of the advanced section? I honestly don't think that's a good idea because its a really solid piece of content and it serves its purpose: a conceptual overview. What's the benefit of removing it? I see more harm than benefit. Maybe we can continue this on slack w/ @rauchg.
